promotion-guide.adoc

```diff
@@ -0,0 +1,137 @@
+
+## Environments
+
+The MRO assumes three "conceptual" environments: DEV, QA and PROD (with short token forms D, Q and P). Those environments are strictly ordered - a state should go from DEV to QA, and then from QA to PROD.
+
+To ensure that software progresses along the DEV -> QA -> PROD path, release bundles from environment DEV can only be installed into QA, and only a release bundle from QA can be installed into PROD. Installing a release bundle from DEV into PROD is not allowed.
+
+Each "conceptual" environment is mapped to an OpenShift namespace:
+
+- DEV to `$PROJECT-dev` (e.g. `foo-dev`)
+- QA to `$PROJECT-test` (e.g. `foo-test`. Note that it is NOT `-qa`!)
+- PROD to `$PROJECT-prod` (e.g. `foo-prod`)
+
+Keep in mind that when you create a new project with OpenDevStack, you get three OpenShift namespaces:
+
+- `foo-dev` (your DEV environment)
+- `foo-test` (your QA environment - unfortunately not named `-qa` for historical reasons)
+- `foo-cd` (where Jenkins runs and the pipelines such as the MRO pipeline are executed)
+
+So while there is a corresponding namespace for DEV and QA, there is no namespace corresponding to the PROD environment out-of-the-box. This is because it is assumed that your PROD environment is likely on another cluster altogether. To create `foo-prod` on another cluster, you (or someone with appropriate rights) can run the script located at https://github.com/opendevstack/ods-core/blob/master/ocp-scripts/create-target-project.sh. Then you need to tell MRO two things: where the API of the external cluster is, and the credentials with which to access it. A typical configuration is:
+
+```
+project: foo
+...
+repositories: [ ... ]
+environments:
+  prod:
+    apiUrl: https://api.example.com
+    credentialsId: foo-cd-foo-prod
+```
+
+This assumes you have the API token credentials stored in a secret of type `kubernetes.io/basic-auth` named `foo-prod` in the `foo-cd` namespace. This secret needs to be synced with Jenkins (which is achieved by labeling it with `credential.sync.jenkins.openshift.io=true`). The stored credentials need to belong to a serviceaccount with rights to admin the `foo-prod` namespace. The easiest way to setup all of this is by running the script located at https://github.com/opendevstack/ods-core/blob/master/ocp-scripts/create-target-sa-secret.sh, which makes use of the output of the `create-target-project.sh` ran earlier.
+
+TIP: It is also possible to have the PROD environment on the same cluster, then you simply create a `foo-prod` namespace next to `foo-dev` and `foo-test`, and allow the `foo-cd:jenkins` account to admin that project. In that case, you do not need to configure anything in `metadata.yml`. The opposite is also possible: you can configure the QA environment to be on a different cluster than the DEV environment - simply follow the instructions above to create a `foo-test` namespace.
```

Comment:
Will there be an update in here https://github.com/opendevstack/ods-core/blob/master/create-projects/create-projects.sh to also create a `$COMPONENT_ID+PROD` namespace? (Could be optional) 
That way newer projects with PROD in the same cluster could avoid the need to create a separate namespace

Reply:
Hmm, right now I don't think so. I don't see it as a bad thing that people need to consider where prod should be. Defaulting it to the same cluster would result in lots of unused namespaces on OpenShift.